src/operator/tensor/elemwise_binary_op.h

```diff
@@ -25,363 +25,938 @@
-  outputs[0].CheckAndAlloc({shape});
-  outputs[1].CheckAndAlloc({shape});
-  MSHADOW_TYPE_SWITCH(outputs[0].dtype(), DType, {
-    MSHADOW_TYPE_SWITCH(outputs[0].aux_type(rowsparse::kIdx), IType, {
-      auto lgrad_idx = outputs[0].aux_data(rowsparse::kIdx).FlatTo1D<xpu, IType>(s);
-      auto rgrad_idx = outputs[1].aux_data(rowsparse::kIdx).FlatTo1D<xpu, IType>(s);
-      auto ograd_idx = inputs[0].aux_data(rowsparse::kIdx).FlatTo1D<xpu, IType>(s);
-      auto lgrad = outputs[0].data().FlatTo1D<xpu, DType>(s);
-      Tensor<xpu, 1, DType> rgrad = outputs[1].data().FlatTo1D<xpu, DType>(s);
-      Tensor<xpu, 1, DType> ograd = inputs[0].data().FlatTo1D<xpu, DType>(s);
-      ASSIGN_DISPATCH(lgrad, req[0], F<LOP>(ograd));
-      ASSIGN_DISPATCH(rgrad, req[1], F<ROP>(ograd));
-      ASSIGN_DISPATCH(lgrad_idx, req[0], F<LOP>(ograd_idx));
-      ASSIGN_DISPATCH(rgrad_idx, req[1], F<ROP>(ograd_idx));
-    });
-  });
-}
-// Only implemented for _backward_add for now
-template<typename xpu, typename LOP, typename ROP>
-void BinaryBackwardUseNoneEx(const nnvm::NodeAttrs& attrs,
-                           const OpContext& ctx,
-                           const std::vector<NDArray>& inputs,
-                           const std::vector<OpReqType>& req,
-                           const std::vector<NDArray>& outputs) {
-  using namespace mshadow;
-  using namespace mshadow::expr;
-  Stream<xpu> *s = ctx.get_stream<xpu>();
-  auto stype = inputs[0].storage_type();
-  CHECK_EQ(stype, kRowSparseStorage) << "Not implemented yet";
-  BinaryBackwardUseNoneRsp<xpu, LOP, ROP>(attrs, ctx, inputs, req, outputs);
-  // TODO(haibin) fallback for kDefaultStorage
-}
-
-template<typename xpu, typename LOP, typename ROP>
-void BinaryBackwardUseNoneWithHalf2(const nnvm::NodeAttrs& attrs,
-                                    const OpContext& ctx,
-                                    const std::vector<TBlob>& inputs,
-                                    const std::vector<OpReqType>& req,
-                                    const std::vector<TBlob>& outputs) {
-  MSHADOW_TYPE_SWITCH_WITH_HALF2(outputs[0].type_flag_, DType, {
-    BinaryBackwardUseNone_<xpu, LOP, ROP, DType>(attrs, ctx, inputs, req, outputs);
-  });
-}
-
-template<typename OP, int Req>
-struct BinaryOpBackwardUseIn {
+
+  /*! \brief Minimum of three */
+  static MSHADOW_XINLINE size_t minthree(const size_t a, const size_t b, const size_t c) {
+    return a < b ? (a < c ? a : c) : (b < c ? b : c);
+  }
+
+  /*! \brief Maximum of three */
+  static MSHADOW_XINLINE size_t maxthree(const size_t a, const size_t b, const size_t c) {
+    return a > b ? (a > c ? a : c) : (b > c ? b : c);
+  }
+
   template<typename DType>
-  MSHADOW_XINLINE static void Map(int i, DType* igrad,
-    const DType* ograd, const DType* lhs, const DType* rhs) {
-    KERNEL_ASSIGN(igrad[i], Req, ograd[i]*OP::Map(lhs[i], rhs[i]));
+  static MSHADOW_XINLINE int LaunchSize(const size_t sz) {
```

Comment:
Is `LaunchSize` used for binary op only? Should we move it to common util file in the future?